File: mobile/android/app/src/main/java/com/volttracker/obdpoc/ObdProbes.java
package com.volttracker.obdpoc;

import java.util.UUID;

/**
 * Static ELM327 connection parameters and the OBD probe command lists used by {@link
 * ObdPollingEngine}. Kept in one place — and package-private so the unit tests can assert these
 * lists stay well-formed.
 */
final class ObdProbes {

    private ObdProbes() {}

    static final UUID ELM327_SPP_UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");
    static final long CONNECT_TIMEOUT_MS = 15000L;
    // After this many consecutive failed reconnects the OBD loop gives up; GPS keeps logging.
    static final int MAX_RECONNECT_ATTEMPTS = 6;
    static final String[] PROTOCOL_PROBES = {"ATSP0", "ATSP6", "ATSP7", "ATSP8"};
    static final String[] CAPABILITY_PROBES = {"0100", "0120", "0140", "0160", "0180", "01A0"};
    // 015B is the standard mode-01 hybrid-battery SOC PID; it answers on the normal bus
    // with no ATSH header, so it is probed alongside the other live-data PIDs.
    static final String[] LIVE_PROBES = {
        "ATRV", "010D", "010C", "0105", "0104", "0111", "0149", "015B", "0142", "011F", "012F",
        "010F", "015C", "01A6"
    };
    // Community-validated Chevy Volt mode-22 PIDs (Volt PID community sheet, see
    // docs/volt-pids-community-sheet.csv). ATSH selects the controller before each group;
    // decode formulas live in ObdProtocol.
    static final String[] VOLT_7E1_PROBES = {
        "222429", // HV pack voltage
        "222414", // HV pack current (signed: discharge positive)
        "222883", // motor A current
        "222884", // motor B current
        "222885", // motor A voltage
        "222886", // motor B voltage
        "222487", // EV distance this cycle
        "222889" // PRNDL / gear selector state
    };
    static final String[] VOLT_7E0_PROBES = {
        "22119F", // engine oil life remaining
        "22119F01", // engine oil life remaining, GM sub-parameter variant
        "221154", // engine oil temperature
        "22203F" // engine torque
    };
    static final String[] VOLT_7E2_PROBES = {
        "221940", // transmission temperature
        "22194001" // transmission temperature, GM sub-parameter variant
    };
    static final String[] VOLT_7E4_PROBES = {
        "22434F", // HV battery temperature
        "224368", // charger AC input voltage
        "224369", // charger AC input current
        "22436B", // charger HV output voltage
        "22436C", // charger HV output current
        "224373", // charging mode
        "22437D", // last charge AC energy
        "2243A5", // HV battery charge count
        "2243AF", // raw high-resolution HV battery SOC
        "224531", // charging level
        "228334", // displayed HV battery SOC
        "2241B2", // battery coolant pump RPM
        "2241B4", // battery coolant valve raw position/state
        "2241B6", // battery coolant heater power
        "22801E", // outside air temperature raw
        "22801F" // outside air temperature filtered
    };
    // Diagnostic-only TPMS discovery candidates. OBDb marks the 7E0 candidates unsupported for the
    // 2018 Volt test data, and public references do not give a reliable universal Volt TPMS PID.
    // Keep these out of live polling until a real scan stores positive responses.
    static final String[] TPMS_7E0_DISCOVERY_PROBES = {
        "22248E", // candidate tire pressure, front-left
        "22248F", // candidate tire pressure, front-right
        "222490", // candidate tire pressure, rear-right
        "222491", // candidate tire pressure, rear-left
        "22C901", // candidate grouped tire pressures
        "22C902" // candidate grouped tire temperatures
    };
    static final String[] TPMS_760_DISCOVERY_PROBES = {
        "224051", // candidate TPMS slot 1
        "224052", // candidate TPMS slot 2
        "224053", // candidate TPMS slot 3
        "224054" // candidate TPMS slot 4
    };
}

File: mobile/android/app/src/main/kotlin/com/volttracker/obdpoc/ObdProbes.kt
package com.volttracker.obdpoc

import java.util.UUID

/**
 * Static ELM327 connection parameters and the OBD probe command lists used by [ObdPollingEngine].
 */
object ObdProbes {
    @JvmField val ELM327_SPP_UUID: UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")
    const val CONNECT_TIMEOUT_MS: Long = 15000L
    const val MAX_RECONNECT_ATTEMPTS: Int = 6

    @JvmField val PROTOCOL_PROBES: Array<String> = arrayOf("ATSP0", "ATSP6", "ATSP7", "ATSP8")

    @JvmField val CAPABILITY_PROBES: Array<String> = arrayOf("0100", "0120", "0140", "0160", "0180", "01A0")

    @JvmField
    val LIVE_PROBES: Array<String> =
        arrayOf(
            "ATRV",
            "010D",
            "010C",
            "0105",
            "0104",
            "0111",
            "0149",
            "015B",
            "0142",
            "011F",
            "012F",
            "010F",
            "015C",
            "01A6",
            "0132",
        )

    @JvmField
    val VOLT_7E1_PROBES: Array<String> =
        arrayOf(
            "222429",
            "222414",
            "222883",
            "222884",
            "222885",
            "222886",
            "222487",
            "222889",
            "221141",
            "221C26",
            "221C28",
            "221C2A",
            "2228CB",
            "22242C",
            "2224B0",
        )

    @JvmField val VOLT_7E0_PROBES: Array<String> = arrayOf("22119F", "22119F01", "221154", "221C47")

    @JvmField val VOLT_7E2_PROBES: Array<String> = arrayOf("221940", "22194001")

    @JvmField
    val VOLT_7E4_PROBES: Array<String> =
        arrayOf(
            "22434F",
            "224368",
            "224369",
            "22436B",
            "22436C",
            "224373",
            "22437D",
            "2243A5",
            "2243AF",
            "2241A3",
            "2245F9",
            "224329",
            "22432A",
            "22432B",
            "22432C",
            "22435F",
            "2240E9",
            "22433B",
            "22433C",
            "224349",
            "22434A",
            "22434B",
            "22434C",
            "221C43",
            "2241A4",
            "22433F",
            "2241B0",
            "22437E",
            "2243A6",
            "2241EC",
            "2241B1",
            "2241B3",
            "2241B5",
            "2282B5",
            "2282B7",
            "224531",
            "228334",
            "2241B2",
            "2241B4",
            "2241B6",
            "22801E",
            "22801F",
        )

    @JvmField val VOLT_7E6_PROBES: Array<String> = arrayOf("224501", "224502")

    @JvmField
    val VOLT_7E7_LAYOUT_PROBES: Array<String> =
        arrayOf(
            "2240D7",
            "2240D9",
            "2240DB",
            "2240DD",
            "2240DF",
            "2240E1",
            "224181",
            "224182",
            "22419F",
            "2241E0",
            "224200",
            "224201",
            "224240",
            "22C218",
            "2240D4",
        )

    // These candidates returned negative/no-data frames on the real car in the June 2026 scan.
    // Keep the executable lists empty until a new TPMS header/addressing source is found.
    @JvmField val TPMS_7E0_DISCOVERY_PROBES: Array<String> = emptyArray()

    @JvmField val TPMS_760_DISCOVERY_PROBES: Array<String> = emptyArray()
}
